KAFKA-9651: Fix ArithmeticException (÷ by 0) in Partitioner impls

tombentley opened the pull request:
This PR fixes KAKFA-9651 by throwing UnknownTopicOrPartitionException instead of ArithmeticException.

tombentley commented:
@cmccabe @mjsax @ewencp please could you review?

guozhangwang reviewed (commented):
@tombentley In the producer code, we always block on metadata for that topic before calling `partition` so it is guaranteed that the cluster would never return empty list.

So the only problem is actually in DefaultStreamPartitioner: since we already have the num.partitions passed in as parameter, we can add a new overload function of `DefaultPartitioner#partition` that instead of taking a Cluster, we just take that num.partitions and then let DefaultStreamPartitioner to call that.

tombentley committed:
KAFKA-9651: Fix ArithmeticException (÷ by 0) in Partitioner impls

This PR fixes KAKFA-9651 by using the number of partitions from the
StreamsMetadataState, rather then from the Cluster.

tombentley commented:
Thanks @guozhangwang, I should have remembered that about the producer.

One thing I don't understand is how using `sourceTopicsInfo.maxPartitions` as the num partitions (as is done by `StreamsMetadataState.getKeyQueryMetadataForKey()`) will result in a non-zero number of partitions, when it appears to me that both code paths will use the same `Cluster` metadata (or maybe I'm wrong about that). 

I updated the commits if you want to take another look.

guozhangwang commented:
`StreamPartitioner#partition`  is used in two places, in the RecordCollectorImpl the caller checks that numPartitions > 0; in the StreamsMetadataState as you asked, it is depending on `clusterMetadata.partitionsForTopic(topic);` indeed, but as long as KafkaStreams is running we are guaranteed that the source topics' metadata are available, and hence the maxPartitions is guaranteed to be non-zero.

tombentley commented:
Thanks for the explanation.

tombentley commented:
@guozhangwang any more comments on this PR?

guozhangwang commented:
test this please

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/DefaultPartitioner.java:
```diff
@@ -50,11 +48,29 @@ public void configure(Map<String, ?> configs) {}
      * @param cluster The current cluster metadata
      */
     public int partition(String topic, Object key, byte[] keyBytes, Object value, byte[] valueBytes, Cluster cluster) {
+        return partition(topic, key, keyBytes, value, valueBytes, cluster, 0);
```
minor: we can just calculate the `numPartitions` in this overload, and then in the other which has `numPartitions` passed in always blindly use that one.

guozhangwang reviewed (commented):
Sorry for the late reply @tombentley . I left a minor comment, otherwise LGTM.

tombentley committed:
Fix review comment

tombentley commented:
Thanks @guozhangwang, now fixed.
